src/main/java/org/opensearch/plugin/insights/core/utils/IndexedPriorityQueue.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.insights.core.utils;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class IndexedPriorityQueue<K, V> {
```

Comment:
Thanks for the changes! We should ideally use IndexedPriorityQueue from a stable source rather than trying to implement it ourselves?

This implementation seems the most popular one used in industry: https://algs4.cs.princeton.edu/24pq/

Reply:
Discussed offline. Seems like the princeton implementation might not be thread safe. Going with this approach for now.